doc/contribute/coding_guidelines/index.rst

```diff
@@ -1255,6 +1255,39 @@ Related GitHub Issues and Pull Requests are tagged with the `Inclusive Language
 .. _CAN in Automation Inclusive Language news post: https://www.can-cia.org/news/archive/view/?tx_news_pi1%5Bnews%5D=699&tx_news_pi1%5Bday%5D=6&tx_news_pi1%5Bmonth%5D=12&tx_news_pi1%5Byear%5D=2020&cHash=784e79eb438141179386cf7c29ed9438
 .. _CAN in Automation Inclusive Language: https://can-newsletter.org/canopen/categories/
 
+
+Rule A.3: Macro name collisions
+===============================
+
+Severity
+--------
+
+Required
+
+Description
+-----------
+
+Commonly used macros such as  ``MIN``, ``MAX``, ``ARRAY_SIZE`` as well as other
+similar ones, must not be modified or protected to avoid name collisions with
+other implementations. In particular, they must not be prefixed to place them in
```

Comment:
> must not be prefixed..

This is ultra-confusing because we already have a large number of `Z_` prefixed macros in our codebase.

What exactly is wrong with changing non-prefixed, zephyr-specific macros to have a zephyr-specific prefix (i.e. a kind of namespace like `Z_` which we use extensively already)?

I don't mean undefining them or guarding them with ifndef.

E.g. since ARRAY_SIZE is used everywhere else in the universe, why not define a Z_ARRAY_SIZE that the Zephyr project uses everywhere to future-proof it?

Reply:
`Z_` (or `z_` for that matter) is supposed to be for Zephyr internal functions.

Reply:
Sure - than another prefix if not that.

Reply:
@carlescufi - can answer the question above for me?

I missed the dev meeting, but just going over the minutes quickly, I've found zero notes justifying why we *should not* "namespace ourselves" (to use the terminology from the minutes). This PR simply says "they must not be prefixed to place them in a Zephyr-specific namespace".

It seems rather non-democratic to try to codify this as policy with no justification other than some people's personal preference.

```
hal: nxp: ARRAY_SIZE collision [Carles]
Issue: https://github.com/zephyrproject-rtos/zephyr/issues/51371
Some HAL modules have their own implementation of common macros like ARRAY_SIZE, MIN, MAX
Three options:
Require modules to namespace
Namespace ourselves
Make HAL use zephyr macros instead
Greg: this is an example of the global namespace problem. There is no way we can force namespaces on downstream projects. Mcuboot is trying integrate with zephyr and with mynewt
David: making HAL use zephyr macros was rejected by the NXP SDK team. We need a process to invoke a wider scope CI run to catch these conflicts
Carles: two modules could collide with each other
Bjarki: we have a lot of defines, none of which are namespaced
David: these macros have been in HAL for years, but we didn’t discover the issue until recently when the include order changed
Kumar: fine for HAL to have the definition and use internally, but poor form to expose externally. User probably expects macros to be defined by OS, not by a HAL
Anas: ifndef violates our coding guidelines. Need header hygiene in the HALs
Carles: agree with Kumar that it is counterintuitive for the HAL to define these common macros instead of the OS. When building for zephyr, the NXP could align to zephyr standards
David: if you want to force users to get Zephyr’s version of these definitions, then you include the zephyr header first
Carles: don’t think it’s future proof to require users to include the zephyr header first. Should implement option 1 or 3 in module
Greg: zephyr isn’t the center of the constellation
Anas: there are at least 15 modules in zephyr today that use ARRAY_SIZE
Carles will send a doc PR
```

I would highly suggest we re-evaluate this.

Reply:
I think that if a prefix is chosen, it should apply to all Zephyr APIs, not just _common macros_. I'd personally have done that from the very beginning, and treat Zephyr as just another library, not the center of the world. But I think it's too late for a change like that, so IMO it's better to just give Zephyr certain naming privileges.

Reply:
> What exactly is wrong with changing non-prefixed, zephyr-specific macros to have a zephyr-specific prefix (i.e. a kind of namespace like Z_ which we use extensively already)?

This was discussed extensively in a couple of meetings and the consensus was to leave the Zephyr ones not namespaced, and namespace the rest.

Reply:
@cfriedt there is no "justification" other than this was the general consensus. The purpose of this PR is to collect objections as well, so feel free to object and, if necessary, we will take it to the TSC for a vote. The info about what was discussed can be found in the different links I posted in the PR description.

Reply:
> @cfriedt there is no "justification" other than this was the general consensus.

The consensus is based on a few justifications:
1. prefixing and namespacing every macro we have is rather disruptive.
2. Even if it is decided to prefix one or two macros, where do we stop?
3. Why does everyone get to use MIN/MAX/ARRAY_SIZE but zephyr?
4. The issue itself is mostly related to header hygene and inclusion of the wrong headers in the wrong places, so prefixing is just a workaround
5. imagine zephyr with a prefix ZEPHYR_MACRO, two modules used to build an app, both define MACRO and both are exposing them publicly, what happens then? Still did not solve the problem.
6. What prefix would chose? Z_ is for private, ZEPHYR_MIN and ZEPHYR_MAX seems rather weird and long, we will spend the next year bikeshedding over a prefix...
7. and there is more....

But as Carles said, this is not "decided", this is how we bring the discussion from the meeting to the wider community. So, objections and other views are welcome.

Reply:
> > @cfriedt there is no "justification" other than this was the general consensus.
> 
> The consensus is based on a few justifications:
> 
> 1. prefixing and namespacing every macro we have is rather disruptive.

Except that it doesn't have to be at all, as I mentioned nearly 2 months ago [here](https://github.com/zephyrproject-rtos/zephyr/pull/50239#issuecomment-1256738002).

The default can be to *enable* defining `ARRAY_SIZE`, `MIN`, and `MAX` to do exactly what they do now. 

e.g.
```
#ifdef CONFIG_DEFINE_ARRAY_SIZE
#define ARRAY_SIZE(x) Z_ARRAY_SIZE(x)
#endif
```

> 2. Even if it is decided to prefix one or two macros, where do we stop?

That's a "slippery slope" logical fallacy. Currently we're talking about 3 macros. Let's stop at 3 for now. If another macro becomes promblematic in the future, then it's 4. Not really a problem.

> 3. Why does everyone get to use MIN/MAX/ARRAY_SIZE but zephyr?

I think this is the root of the issue - it's inconvenient. People don't want to change their habits. That's understandable.

> 4. The issue itself is mostly related to header hygene and inclusion of the wrong headers in the wrong places, so prefixing is just a workaround

Incorrect - prefixing (i.e. namespacing) is to explicitly avoid collisions and do so in a way that is sustainable. It's the same reason we prefix virtually all public kernel API with `K_` or `k_`. 

Simply do not guard the definition with `#ifndef` and then we don't have include order problems, but give end-users the option (via Kconfig) to opt-out.

> 5. imagine zephyr with a prefix ZEPHYR_MACRO, two modules used to build an app, both define MACRO and both are exposing them publicly, what happens then? Still did not solve the problem.
> 6. What prefix would chose? Z_ is for private, ZEPHYR_MIN and ZEPHYR_MAX seems rather weird and long, we will spend the next year bikeshedding over a prefix...
> 7. and there is more....

Reply:
I missed the dev meeting, and sorry for coming off sharp. I personally don't care about this too much.. I just wanted to highlight that there is a better way of doing things technically, but I won't block this anymore. No need to vote on it in the TSC meeting.

Reply:
> Is "must" really the right word to use? Maybe "should"?

yeah, maybe should, or if this  " must be implemented, and its implementation verified.  Period! " then *shall*. 

And btw, we need to be consistent with this language across our docs, should, shall and must, *shall* have the same meaning across all zephyr docs, or it will be confusing.